src/rng.rs

```diff
@@ -0,0 +1,19 @@
+//! Random Number Generator Interface
+
+use generic_array::{ArrayLength, GenericArray};
+use nb;
+
+/// Nonblocking stream of random bytes.
+// reason: No implementation or users yet
+pub trait Read<N>
```

Comment:
I actually tried with an associated type first, but did it in the wrong way (I tried to use `GenericArray` rather than `ArrayLength` as the bound, which obviously won't work).

I think I'm actually in favor of the associated type method -- I don't see enough value in multiple implementations for the same type to justify the ergonomic penalty. I'll defer to whatever more experienced Rust devs have to say, though.

```Rust
/// Nonblocking stream of random bytes.
// reason: No implementation or users yet
pub trait Read {
    /// Number of bytes in a result
    type Size: ArrayLength<u8>;

    /// An enumeration of RNG errors.
    ///
    /// For infallible implementations, will be `Void`
    type Error;

    /// Get a number of bytes from the RNG.
    fn read(&mut self) -> nb::Result<GenericArray<u8, Self::Size>, Self::Error>;
}
```

Reply:
So, I thought some more about this. Let's take it step by step.

First, I've come to the conclusion that the type argument makes no sense here:
1. A consumer of this API that values portability should always choose a size of one, because larger sizes might not be available on a given platform.
2. That means that any implementation that can provide more bytes in one go also needs to provide implementations for fewer bytes. This seems tedious, and sounds like something an implementer might forget.

So, if any implementation needs to provide a trait implementation for one byte, and no consumer should expect an implementation to deliver more than one byte (unless they _absolutely_ need more than one byte at a time, but when would that be the case?), why are we bothering with a type argument at all, and not just use `u8`?

But let's say we find `u8` too limiting in the general case and want to use the hardware to its full extent. So an associated type it is. But:
1. As a consumer of the API, we don't know what this associated type will be, unless we specify it in a where clause. That would lead to the same problems the type argument had, except now it's worse, because implementations capable of providing more than one byte can't provide the 1-byte implementation as a fallback.
2. So we accept that we don't know what the type is and get back a `GenericArray`. What is the only thing we can do with that? Get a slice. What happens if that slice contains fewer bytes than we need? We note down how many bytes we got, then call again, until we have enough bytes.

This is very inconvenient, which is exactly the criticism that the initial proposal received. Except no, this time it's worse. At least in the initial proposal, the RNG put the bytes into our buffer. Now we have to maintain a buffer, and copy bytes out `GenericArrays` into it ourselves.

So... am I missing some deeper thought that went into this iteration of the proposal? Because if not, the initial proposal looks way more attractive to me.

Reply:
Just to confirm, was the the original proposal you're referring to this?

```Rust
// rng.rs
pub trait Read {
    type Error;

    fn read(&mut self) -> nb::Result(u8, Self::Error);
}
```

I don't think you're missing any deeper reasoning, and I see your point entirely. I guess it's more-or-less premature optimization to resist throwing away extra data that's provided by some TRNGs.

I'm happy to roll back my PR to the above version. If nothing else, it avoids leaking hardware details through the trait.

One method that I tried to propose (but that I did in an unclear way) would be for the trait implementor itself to hold on to the extra bytes, and feed them out one at a time as they're requested. The obvious question becomes: where does that state get stored? My proposal is to store it in `self`.

It already seems like the accepted pattern to wrap register-blocks in new structural types (a-la `stm32f30x_hal::i2c:I2c`, stm32f30x_hal::spi::Spi`, etc). Trait-users already expect these implementations to be stateful -- in every case, there's implicit state stored in hardware registers.

On the other hand, I've heard it said elsewhere (though I can't immediately find it) that `embedded-hal` trait implementations must _only_ depend on the state in hardware registers. I'll admit that I don't entirely understand the logic behind this restriction. It also seems like there are crates that break this -- [`shared-bus`](https://crates.io/crates/shared-bus) being the first that comes to mind, since it potentially depends on an in-memory mutex.

I don't see this as making or breaking the case for a non-generic trait, but if HAL implementors at least had the _opportunity_ for this kind of optimization, then there's absolutely no argument to be made for the generic, variably-sized variant.

I feel like I've let this response ramble a little much, so here is the core of my thoughts:

- The only thing we lose by simply returning a `u8` is, as you say, "full use of the hardware." I agree that that's not worth it
- There's a potential implementation strategy which _does_ make full use of the hardware, while relying on storing non-register state in `self`

Reply:
> Just to confirm, was the the original proposal you're referring to this?

Actually, I was referring to what @Nemo157 proposed:
``` rust
fn read(&mut self, buf: &mut [u8]) -> nb::Result<usize, Self::Error>
```
Sorry for referring to it as the initial proposal. For some reason I thought it was. At this point, this slice-based version is the one I prefer. It's simple, takes full advantage of hardware capabilities, and it's straight-forward to write a (blocking or non-blocking) `read_all` wrapper, that provides as many bytes as the user wants.

Of course, I might be missing something, or might have overseen something from the discussion here, so I'm open to being convinced of something else.

Regarding your 1-byte version:
``` rust
fn read(&mut self) -> nb::Result<u8, Self::Error>;
```

I think this is fine, too. Not my preferred version, but a good choice.

> One method that I tried to propose (but that I did in an unclear way) would be for the trait implementor itself to hold on to the extra bytes, and feed them out one at a time as they're requested.
> [...]
> On the other hand, I've heard it said elsewhere (though I can't immediately find it) that `embedded-hal` trait implementations must only depend on the state in hardware registers. I'll admit that I don't entirely understand the logic behind this restriction.

I think this approach is fine. I don't see why `embedded-hal` trait implementations shouldn't have internal (non-hardware) state. Maybe there were good reasons for that statement in its original context, but it certainly doesn't hold in the general case.

> I don't see this as making or breaking the case for a non-generic trait, but if HAL implementors at least had the opportunity for this kind of optimization, then there's absolutely no argument to be made for the generic, variably-sized variant.

Agreed.

To summarize my own position:
- I currently prefer the slice-based variant, for the reasons stated above.
- I find the one-byte variant to be a good option, too. I think caching bytes is a fine implementation strategy.
- I think this whole generic argument discussion (either via type argument or associated type) went entirely in the wrong direction.

Reply:
> On the other hand, I've heard it said elsewhere (though I can't immediately find it) that embedded-hal trait implementations must only depend on the state in hardware registers.

I think that may have been my statement but in a different context: Trait implementations should not rely on internal state representing physical state, e.g. that GPIOs are and stay in the physical state they were set to by an implementation. Buffering data in internal state should be fine, though we shouldn't overdo it...